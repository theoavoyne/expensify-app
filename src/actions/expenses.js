import database from '../firebase/firebase';

// ADD EXPENSE

export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => (
  (dispatch) => {
    const expense = {
      description: expenseData.description || '',
      note: expenseData.note || '',
      amount: expenseData.amount || 0,
      createdAt: expenseData.createdAt || 0
    };

    return database
      .ref('expenses')
      .push(expense)
      .then((ref) => {
        dispatch(addExpense({
          id: ref.key,
          ...expense
        }));
      });
  }
);

// REMOVE EXPENSE

export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

// EDIT EXPENSE

export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

// SET EXPENSES

export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

export const startSetExpenses = () => (
  (dispatch) => (
    database
      .ref('expenses')
      .once('value')
      .then((snapshot) => {
        const expensesObj = snapshot.val() || {};

        const expenses = Object.keys(expensesObj).map((key) => ({
          id: key,
          ...expensesObj[key]
        }));

        dispatch(setExpenses(expenses));
      })
  )
);
