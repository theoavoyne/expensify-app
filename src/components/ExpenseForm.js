import React from 'react';
import DatePicker from 'react-datepicker';
import moment from 'moment';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      error: ''
    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  }

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  }

  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  }

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt: moment(createdAt) }));
    }
  }

  onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Please provide description and amount.' }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  }

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form action="" onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <DatePicker
            selected={this.state.createdAt.toDate()}
            onChange={this.onDateChange}
          />
          <textarea
            placeholder="Add a note for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          />
          <button type="submit">Add Expense</button>
        </form>
      </div>
    );
  }
}
