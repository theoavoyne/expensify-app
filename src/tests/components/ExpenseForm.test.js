import React from 'react';
import { shallow } from 'enzyme';
import moment from 'moment';
import _ from 'lodash';

import ExpenseForm from '../../components/ExpenseForm';
import expenses from '../fixtures/expenses';

test('Should render ExpenseForm correctly', () => {
  const wrapper = shallow(<ExpenseForm />);
  expect(wrapper).toMatchSnapshot();
});

test('Should render ExpenseForm with expense data correctly', () => {
  const wrapper = shallow(<ExpenseForm expense={expenses[1]} />);
  expect(wrapper).toMatchSnapshot();
});

test('Should render error for invalid form submission', () => {
  const wrapper = shallow(<ExpenseForm />);
  const event = { preventDefault: () => {} };
  wrapper.find('form').simulate('submit', event);
  expect(wrapper.state('error')).toBeTruthy();
  expect(wrapper).toMatchSnapshot();
});

test('Should set description on input change', () => {
  const wrapper = shallow(<ExpenseForm />);
  const value = 'New description';
  wrapper.find('input').at(0).simulate('change', {
    target: { value }
  });
  expect(wrapper.state('description')).toBe(value);
});

test('Should set note on on textarea change', () => {
  const wrapper = shallow(<ExpenseForm />);
  const value = 'New note';
  wrapper.find('textarea').simulate('change', {
    target: { value }
  });
  expect(wrapper.state('note')).toBe(value);
});

test('Should set amount if valid input', () => {
  const wrapper = shallow(<ExpenseForm />);
  const value = '23.50';
  wrapper.find('input').at(1).simulate('change', {
    target: { value }
  });
  expect(wrapper.state('amount')).toBe(value);
});

test('Should not set amount if invalid input', () => {
  const wrapper = shallow(<ExpenseForm />);
  const value = '12.122';
  wrapper.find('input').at(1).simulate('change', {
    target: { value }
  });
  expect(wrapper.state('amount')).toBeFalsy();
});

test('Should call onSubmit prop for valid form submission', () => {
  const onFormSubmit = jest.fn();
  const wrapper = shallow(
    <ExpenseForm
      onSubmit={onFormSubmit}
      expense={expenses[1]}
    />
  );
  const event = { preventDefault: () => {} };
  wrapper.find('form').simulate('submit', event);
  expect(wrapper.state('error')).toBe('');
  expect(onFormSubmit).toHaveBeenLastCalledWith(
    _.omit(expenses[1], 'id')
  );
});

test('Should set new date on date change', () => {
  const wrapper = shallow(<ExpenseForm />);
  const date = moment(1000);
  wrapper.find('t').prop('onChange')(date.valueOf());
  expect(wrapper.state('createdAt')).toEqual(date);
});
