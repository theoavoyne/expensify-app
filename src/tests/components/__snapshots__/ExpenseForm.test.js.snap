// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseForm correctly 1`] = `
<div>
  <form
    action=""
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Amount"
      type="text"
      value=""
    />
    <t
      allowSameDay={false}
      dateFormat="MM/dd/yyyy"
      dateFormatCalendar="LLLL yyyy"
      disabled={false}
      disabledKeyboardNavigation={false}
      dropdownMode="scroll"
      inlineFocusSelectedMonth={false}
      monthsShown={1}
      nextMonthButtonLabel="Next Month"
      nextYearButtonLabel="Next Year"
      onBlur={[Function]}
      onChange={[Function]}
      onClickOutside={[Function]}
      onFocus={[Function]}
      onInputClick={[Function]}
      onInputError={[Function]}
      onKeyDown={[Function]}
      onMonthChange={[Function]}
      onSelect={[Function]}
      onYearChange={[Function]}
      preventOpenOnFocus={false}
      previousMonthButtonLabel="Previous Month"
      previousYearButtonLabel="Previous Year"
      readOnly={false}
      renderDayContents={[Function]}
      selected={1970-01-01T00:00:00.000Z}
      shouldCloseOnSelect={true}
      showMonthYearPicker={false}
      showQuarterYearPicker={false}
      showTimeInput={false}
      showTimeSelect={false}
      strictParsing={false}
      timeCaption="Time"
      timeInputLabel="Time"
      timeIntervals={30}
      withPortal={false}
    />
    <textarea
      onChange={[Function]}
      placeholder="Add a note for your expense (optional)"
      value=""
    />
    <button
      type="submit"
    >
      Add Expense
    </button>
  </form>
</div>
`;

exports[`Should render ExpenseForm with expense data 1`] = `
<div>
  <form
    action=""
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value="Rent"
    />
    <input
      onChange={[Function]}
      placeholder="Amount"
      type="text"
      value="1095"
    />
    <t
      allowSameDay={false}
      dateFormat="MM/dd/yyyy"
      dateFormatCalendar="LLLL yyyy"
      disabled={false}
      disabledKeyboardNavigation={false}
      dropdownMode="scroll"
      inlineFocusSelectedMonth={false}
      monthsShown={1}
      nextMonthButtonLabel="Next Month"
      nextYearButtonLabel="Next Year"
      onBlur={[Function]}
      onChange={[Function]}
      onClickOutside={[Function]}
      onFocus={[Function]}
      onInputClick={[Function]}
      onInputError={[Function]}
      onKeyDown={[Function]}
      onMonthChange={[Function]}
      onSelect={[Function]}
      onYearChange={[Function]}
      preventOpenOnFocus={false}
      previousMonthButtonLabel="Previous Month"
      previousYearButtonLabel="Previous Year"
      readOnly={false}
      renderDayContents={[Function]}
      selected={1969-12-28T00:00:00.000Z}
      shouldCloseOnSelect={true}
      showMonthYearPicker={false}
      showQuarterYearPicker={false}
      showTimeInput={false}
      showTimeSelect={false}
      strictParsing={false}
      timeCaption="Time"
      timeInputLabel="Time"
      timeIntervals={30}
      withPortal={false}
    />
    <textarea
      onChange={[Function]}
      placeholder="Add a note for your expense (optional)"
      value=""
    />
    <button
      type="submit"
    >
      Add Expense
    </button>
  </form>
</div>
`;

exports[`Should render error for invalid form submission 1`] = `
<div>
  <form
    action=""
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Amount"
      type="text"
      value=""
    />
    <t
      allowSameDay={false}
      dateFormat="MM/dd/yyyy"
      dateFormatCalendar="LLLL yyyy"
      disabled={false}
      disabledKeyboardNavigation={false}
      dropdownMode="scroll"
      inlineFocusSelectedMonth={false}
      monthsShown={1}
      nextMonthButtonLabel="Next Month"
      nextYearButtonLabel="Next Year"
      onBlur={[Function]}
      onChange={[Function]}
      onClickOutside={[Function]}
      onFocus={[Function]}
      onInputClick={[Function]}
      onInputError={[Function]}
      onKeyDown={[Function]}
      onMonthChange={[Function]}
      onSelect={[Function]}
      onYearChange={[Function]}
      preventOpenOnFocus={false}
      previousMonthButtonLabel="Previous Month"
      previousYearButtonLabel="Previous Year"
      readOnly={false}
      renderDayContents={[Function]}
      selected={1970-01-01T00:00:00.000Z}
      shouldCloseOnSelect={true}
      showMonthYearPicker={false}
      showQuarterYearPicker={false}
      showTimeInput={false}
      showTimeSelect={false}
      strictParsing={false}
      timeCaption="Time"
      timeInputLabel="Time"
      timeIntervals={30}
      withPortal={false}
    />
    <textarea
      onChange={[Function]}
      placeholder="Add a note for your expense (optional)"
      value=""
    />
    <button
      type="submit"
    >
      Add Expense
    </button>
  </form>
</div>
`;

exports[`Should render error for invalid form submission 2`] = `
<div>
  <p>
    Please provide description and amount.
  </p>
  <form
    action=""
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Amount"
      type="text"
      value=""
    />
    <t
      allowSameDay={false}
      dateFormat="MM/dd/yyyy"
      dateFormatCalendar="LLLL yyyy"
      disabled={false}
      disabledKeyboardNavigation={false}
      dropdownMode="scroll"
      inlineFocusSelectedMonth={false}
      monthsShown={1}
      nextMonthButtonLabel="Next Month"
      nextYearButtonLabel="Next Year"
      onBlur={[Function]}
      onChange={[Function]}
      onClickOutside={[Function]}
      onFocus={[Function]}
      onInputClick={[Function]}
      onInputError={[Function]}
      onKeyDown={[Function]}
      onMonthChange={[Function]}
      onSelect={[Function]}
      onYearChange={[Function]}
      preventOpenOnFocus={false}
      previousMonthButtonLabel="Previous Month"
      previousYearButtonLabel="Previous Year"
      readOnly={false}
      renderDayContents={[Function]}
      selected={1970-01-01T00:00:00.000Z}
      shouldCloseOnSelect={true}
      showMonthYearPicker={false}
      showQuarterYearPicker={false}
      showTimeInput={false}
      showTimeSelect={false}
      strictParsing={false}
      timeCaption="Time"
      timeInputLabel="Time"
      timeIntervals={30}
      withPortal={false}
    />
    <textarea
      onChange={[Function]}
      placeholder="Add a note for your expense (optional)"
      value=""
    />
    <button
      type="submit"
    >
      Add Expense
    </button>
  </form>
</div>
`;
